Warning with LLVM 12

<!--

NOTE: this form is for bug reports only. For questions or troubleshooting, please see the protobuf mailing list: https://groups.google.com/forum/#!forum/protobuf

-->

**What version of protobuf and what language are you using?**
Version: 3.17.0-rc1
Language: N/A (compiler)

**What operating system (Linux, Windows, ...) and version?** Linux (Archlinux)

**What runtime / compiler are you using (e.g., python version or gcc version)** LLVM 12.0.0

**What did you do?**
Tried to compile `protoc` through the Bazel rules

**What did you expect to see**
A successful build of protoc

**What did you see instead?**
A failure due to a warning:
```
external/com_google_protobuf/src/google/protobuf/descriptor.cc:1038:9: error: passing variable 'unknown_enum_values_by_number_' by reference requires holding mutex 'unknown_enum_values_mu_' [-Werror,-Wthread-safety-reference]
        unknown_enum_values_by_number_, std::make_pair(parent, number));
        ^
external/com_google_protobuf/src/google/protobuf/descriptor.cc:1048:9: error: passing variable 'unknown_enum_values_by_number_' by reference requires holding mutex 'unknown_enum_values_mu_' [-Werror,-Wthread-safety-reference]
        unknown_enum_values_by_number_, std::make_pair(parent, number));
        ^
2 errors generated.

```

The same with 3.17.0 on macOS, Bazel 4.1.0rc4:

```
INFO: From Compiling src/google/protobuf/descriptor.cc:
external/com_google_protobuf/src/google/protobuf/descriptor.cc:1038:9: warning: passing variable 'unknown_enum_values_by_number_' by reference requires holding mutex 'unknown_enum_values_mu_' [-Wthread-safety-reference]
        unknown_enum_values_by_number_, std::make_pair(parent, number));
        ^
external/com_google_protobuf/src/google/protobuf/descriptor.cc:1048:9: warning: passing variable 'unknown_enum_values_by_number_' by reference requires holding mutex 'unknown_enum_values_mu_' [-Wthread-safety-reference]
        unknown_enum_values_by_number_, std::make_pair(parent, number));
        ^
2 warnings generated.
INFO: From Compiling src/google/protobuf/descriptor.cc [for host]:
external/com_google_protobuf/src/google/protobuf/descriptor.cc:1038:9: warning: passing variable 'unknown_enum_values_by_number_' by reference requires holding mutex 'unknown_enum_values_mu_' [-Wthread-safety-reference]
        unknown_enum_values_by_number_, std::make_pair(parent, number));
        ^
external/com_google_protobuf/src/google/protobuf/descriptor.cc:1048:9: warning: passing variable 'unknown_enum_values_by_number_' by reference requires holding mutex 'unknown_enum_values_mu_' [-Wthread-safety-reference]
        unknown_enum_values_by_number_, std::make_pair(parent, number));
        ^
2 warnings generated.
```

--- a/src/google/protobuf/stubs/mutex.h
+++ b/src/google/protobuf/stubs/mutex.h
@@ -52,10 +52,12 @@
   __attribute__((acquire_capability(__VA_ARGS__)))
 #define GOOGLE_PROTOBUF_RELEASE(...) \
   __attribute__((release_capability(__VA_ARGS__)))
+#define GOOGLE_PROTOBUF_SCOPED_CAPABILITY __attribute__((scoped_lockable))
 #define GOOGLE_PROTOBUF_CAPABILITY(x) __attribute__((capability(x)))
 #else
 #define GOOGLE_PROTOBUF_ACQUIRE(...)
 #define GOOGLE_PROTOBUF_RELEASE(...)
+#define GOOGLE_PROTOBUF_SCOPED_CAPABILITY
 #define GOOGLE_PROTOBUF_CAPABILITY(x)
 #endif
 
@@ -140,10 +142,12 @@ class GOOGLE_PROTOBUF_CAPABILITY("mutex") PROTOBUF_EXPORT WrappedMutex {
 using Mutex = WrappedMutex;
 
 // MutexLock(mu) acquires mu when constructed and releases it when destroyed.
-class PROTOBUF_EXPORT MutexLock {
+class GOOGLE_PROTOBUF_SCOPED_CAPABILITY PROTOBUF_EXPORT MutexLock {
  public:
-  explicit MutexLock(Mutex *mu) : mu_(mu) { this->mu_->Lock(); }
-  ~MutexLock() { this->mu_->Unlock(); }
+  explicit MutexLock(Mutex *mu) GOOGLE_PROTOBUF_ACQUIRE(mu) : mu_(mu) {
+    this->mu_->Lock();
+  }
+  ~MutexLock() GOOGLE_PROTOBUF_RELEASE() { this->mu_->Unlock(); }
  private:
   Mutex *const mu_;
   GOOGLE_DISALLOW_EVIL_CONSTRUCTORS(MutexLock);

